src/CoreFoundation/CFSocket.cs

```diff
@@ -300,90 +355,68 @@ public CFSocket (AddressFamily family, SocketType type, ProtocolType proto)
 
 		public CFSocket (AddressFamily family, SocketType type, ProtocolType proto, CFRunLoop loop)
 			: this (CFSocketSignature.AddressFamilyToInt (family),
-			        CFSocketSignature.SocketTypeToInt (type),
-			        CFSocketSignature.ProtocolToInt (proto), loop)
+				CFSocketSignature.SocketTypeToInt (type),
+				CFSocketSignature.ProtocolToInt (proto), loop)
 		{
 		}
 
+		const CFSocketCallBackType defaultCallbackTypes = CFSocketCallBackType.DataCallBack | CFSocketCallBackType.ConnectCallBack;
 		CFSocket (int family, int type, int proto, CFRunLoop loop)
 		{
-			var cbTypes = CFSocketCallBackType.DataCallBack | CFSocketCallBackType.ConnectCallBack;
-
-			gch = GCHandle.Alloc (this);
-			var ctx = new CFStreamClientContext ();
-			ctx.Info = GCHandle.ToIntPtr (gch);
-
-			var ptr = Marshal.AllocHGlobal (Marshal.SizeOf (typeof(CFStreamClientContext)));
-			try {
-				Marshal.StructureToPtr (ctx, ptr, false);
-				handle = CFSocketCreate (
-					IntPtr.Zero, family, type, proto, (nuint) (ulong) cbTypes, OnCallback, ptr);
-			} finally {
-				Marshal.FreeHGlobal (ptr);
-			}
-
-			if (handle == IntPtr.Zero)
-				throw new CFSocketException (CFSocketError.Error);
-			gch = GCHandle.Alloc (this);
-
-			var source = new CFRunLoopSource (CFSocketCreateRunLoopSource (IntPtr.Zero, handle, 0));
-			loop.AddSource (source, CFRunLoop.ModeDefault);
+			var ctx = new CFSocketContext (this);
```

Comment:
You can move the creation of the `CFSocketContext` into the `Initialize` method as well, it's identical in all three cases.

Reply:
How? We depend on the handle form the PInvoke. And the pinvoke depends on the struct.

Reply:
Ah duh, nvm.